### DISCUSSION TITLE
[Issues] getInitialProps in _app.tsx

### DISCUSSION THREAD

**Discussion by mtr1990** [vercel/next.js#31447] [Help]
### What version of Next.js are you using?

"next": "^12.0.3",

### What version of Node.js are you using?

v14.17.3

### What browser are you using?

Chrome

### What operating system are you using?

macOS

### How are you deploying your application?

Vercel

### Describe the Bug

For the special case, I want to use `getInitialProps` in _app.tsx

**When I build I get this error:** `Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps``

### Expected Behavior

I have researched but still haven't found an alternative to `getInitialProps` in  _app.tsx

If it is handled on the page, this is quite simple

but I want to pass props to Context and this only works with `getInitialProps`

Is there any alternative solution that solves this case? Please help.

### To Reproduce

`Warning: You have opted-out of Automatic Static Optimization due to `getInitialProps` in `pages/_app`. This does not opt-out pages with `getStaticProps`

```js
export default function MyApp(props: MyAppProps) {
  const { Component, pageProps, settings } = props;

  return (
    <>
      <Head>
        <meta name="viewport" content="initial-scale=1, width=device-width" />
      </Head>
      <SettingsProvider defaultSettings={settings}> // => Need to pass props in this context
        <Layout>
          <Component {...pageProps} />
        </Layout>
      </SettingsProvider>
    </>
  );
}

```

```js
MyApp.getInitialProps = async (context: AppContext) => {
  const appProps = await App.getInitialProps(context);

  const cookies = cookie.parse(
    context.ctx.req ? context.ctx.req.headers.cookie || '' : document.cookie
  );

  const settings = getSettings(cookies);

  return {
    ...appProps,
    settings,
  };
};

```

**Comment by maciusio** [comment]
I'm not sure if I understood your issue correctly. First of all the message
`**Warning**: You have opted-out of Automatic Static Optimization due to getInitialProps in pages/_app. This does not opt-out pages with getStaticProps`
is not an error, it's a warning, and that's a big difference.

If you want to have some common logic to get some blocking data for every single page in your application then getInitialProps in custom _app is a perfectly fine place as described here https://nextjs.org/docs/advanced-features/custom-app. There is a small bug in the docs `This disables the ability to perform automatic static optimization, causing every page in your app to be server-side rendered.`. That is not true. What the warning message says is true.

So, what is the warning about? So basically if in your page's code you have `getServerSideProps` then you can safely ignore the warning; it won't affect you. If you have in your page's code `getStaticProps` then you can ignore the warning as well. 

The only case the warning is warning your :) is when a page has neither `getServerSideProps` nor `getStaticSideProps`. Without `getInitialProps` in your `_app` such a page would be optimised and treated as a page that can be pre-rendered (statically build). With `getInitialProps` in `_app` such a page will be treated as a page that should be server-side rendered (per request).
